modules/affile/affile-source.c

```diff
@@ -330,95 +331,149 @@ affile_sd_queue(LogPipe *s, LogMessage *msg, const LogPathOptions *path_options,
+                evt_tag_int("fd", fd),
+                NULL);
+      log_pipe_unref((LogPipe *) self->reader);
+      self->reader = NULL;
+      close(fd);
+      return FALSE;
+    }
 
-      poll_events = affile_sd_construct_poll_events(self, fd);
-      if (!poll_events)
-        {
-          close(fd);
-          return FALSE;
-        }
+  affile_sd_recover_state(s, cfg, proto);
+
+  return TRUE;
+}
 
-      proto = affile_sd_construct_proto(self, fd);
-      self->reader = log_reader_new(self->super.super.super.cfg);
-      log_reader_reopen(self->reader, proto, poll_events);
-
-      log_reader_set_options(self->reader,
-                             s,
-                             &self->reader_options,
-                             STATS_LEVEL1,
-                             SCS_FILE,
-                             self->super.super.id,
-                             self->filename->str);
-      /* NOTE: if the file could not be opened, we ignore the last
-       * remembered file position, if the file is created in the future
-       * we're going to read from the start. */
-      
-      log_pipe_append((LogPipe *) self->reader, s);
-      if (!log_pipe_init((LogPipe *) self->reader))
+static gboolean
+_open_file(LogPipe *s, GlobalConfig *cfg)
+{
+  AFFileSourceDriver *self = (AFFileSourceDriver *) s;
+  gint fd;
+
+  if (!affile_sd_open_file(self, self->filename->str, &fd))
+    {
+      if (self->follow_freq == 0)
         {
-          msg_error("Error initializing log_reader, closing fd",
-                    evt_tag_int("fd", fd),
+          msg_error("Error opening file for reading",
+                    evt_tag_str("filename", self->filename->str),
+                    evt_tag_errno(EVT_TAG_OSERROR, errno),
                     NULL);
-          log_pipe_unref((LogPipe *) self->reader);
-          self->reader = NULL;
-          close(fd);
-          return FALSE;
+          return self->super.super.optional;
         }
-      affile_sd_recover_state(s, cfg, proto);
-    }
-  else
-    {
-      msg_error("Error opening file for reading",
-                evt_tag_str("filename", self->filename->str),
-                evt_tag_errno(EVT_TAG_OSERROR, errno),
-                NULL);
-      return self->super.super.optional;
+
+      msg_info("Follow-mode file source not found, deferring open",
+               evt_tag_str("filename", self->filename->str),
+               NULL);
+      fd = -1;
     }
-  return TRUE;
 
+  return _reopen_log_reader(s, cfg, fd);
+}
+
+static gboolean
+affile_sd_init(LogPipe *s)
+{
+  AFFileSourceDriver *self = (AFFileSourceDriver *) s;
+  GlobalConfig *cfg = log_pipe_get_config(s);
+
+  if (!log_src_driver_init_method(s))
+    return FALSE;
+
+  log_reader_options_init(&self->reader_options, cfg, self->super.super.group);
+
+  if (!_check_multiline_options(self))
+    return FALSE;
+
+  self->reader = cfg_persist_config_fetch(cfg, affile_sd_format_persist_name(self));
+  if (self->reader)
```

Comment:
this is where we could create a new LogReader instance in case cfg_persist_config_fetch() returns NULL.

see my comments around _configure_logreader_...() function